### ISSUE TITLE
Retired nodes becoming followers

### ISSUE THREAD

**Issue by wintersteiger** [microsoft/CCF#2456]
The rotation_test (and likely others) frequently results in nodes changing from the retired state to the follower state. In the log files I see things like these:

```
... $ for i in workspace/rotation_test_*/out; do echo $i; grep Becoming $i; done
workspace/rotation_test_0/out
2021-04-12T14:18:34.197290Z -0.091 0   [info ] ../src/consensus/aft/raft.h:2347     | Becoming leader 8ea41e38682cfd5c87528bef3337c7f08464f93e68ffce423e8ba79edb3bfc33: 2
2021-04-12T14:18:45.575498Z -0.003 0   [info ] ../src/consensus/aft/raft.h:2398     | Becoming retired 8ea41e38682cfd5c87528bef3337c7f08464f93e68ffce423e8ba79edb3bfc33: 2
2021-04-12T14:18:47.943018Z -0.002 0   [info ] ../src/consensus/aft/raft.h:2384     | Becoming follower 8ea41e38682cfd5c87528bef3337c7f08464f93e68ffce423e8ba79edb3bfc33: 3
workspace/rotation_test_1/out
2021-04-12T14:18:36.278080Z -0.002 0   [info ] ../src/consensus/aft/raft.h:2384     | Becoming follower d8c06b7581f77496f9c46e1b3fc76c0d4fb9e819e27beb12a4d70585fd5ab6ff: 2
2021-04-12T14:18:47.915849Z -0.002 0   [info ] ../src/consensus/aft/raft.h:2307     | Becoming candidate d8c06b7581f77496f9c46e1b3fc76c0d4fb9e819e27beb12a4d70585fd5ab6ff: 3
2021-04-12T14:18:48.076984Z -0.002 0   [info ] ../src/consensus/aft/raft.h:2347     | Becoming leader d8c06b7581f77496f9c46e1b3fc76c0d4fb9e819e27beb12a4d70585fd5ab6ff: 3
2021-04-12T14:18:55.822770Z -0.058 0   [info ] ../src/consensus/aft/raft.h:2398     | Becoming retired d8c06b7581f77496f9c46e1b3fc76c0d4fb9e819e27beb12a4d70585fd5ab6ff: 3
2021-04-12T14:18:59.044514Z -0.002 0   [info ] ../src/consensus/aft/raft.h:2384     | Becoming follower d8c06b7581f77496f9c46e1b3fc76c0d4fb9e819e27beb12a4d70585fd5ab6ff: 4
,,,
workspace/rotation_test_3/out
2021-04-12T14:18:54.888878Z -0.007 0   [info ] ../src/consensus/aft/raft.h:2384     | Becoming follower 1efdf2c85f19c26c0351b7881b272b87b152820e3cd822a2d9d020f81f1d0374: 3
2021-04-12T14:18:59.032369Z -0.001 0   [info ] ../src/consensus/aft/raft.h:2307     | Becoming candidate 1efdf2c85f19c26c0351b7881b272b87b152820e3cd822a2d9d020f81f1d0374: 4
2021-04-12T14:18:59.124822Z -0.003 0   [info ] ../src/consensus/aft/raft.h:2347     | Becoming leader 1efdf2c85f19c26c0351b7881b272b87b152820e3cd822a2d9d020f81f1d0374: 4
2021-04-12T14:19:08.288392Z -0.002 0   [info ] ../src/consensus/aft/raft.h:2398     | Becoming retired 1efdf2c85f19c26c0351b7881b272b87b152820e3cd822a2d9d020f81f1d0374: 4
2021-04-12T14:19:11.651050Z -0.003 0   [info ] ../src/consensus/aft/raft.h:2384     | Becoming follower 1efdf2c85f19c26c0351b7881b272b87b152820e3cd822a2d9d020f81f1d0374: 5
,,,
```

It looks like this happens when a retired node receives a vote request:

```
...
2021-04-12T14:18:55.822770Z -0.058 0   [info ] ../src/consensus/aft/raft.h:2398     | Becoming retired d8c06b7581f77496f9c46e1b3fc76c0d4fb9e819e27beb12a4d70585fd5ab6ff: 3
...
2021-04-12T14:18:59.043947Z -0.002 0   [info ] ../src/node/channels.h:793           | Node channel with 1efdf2c85f19c26c0351b7881b272b87b152820e3cd822a2d9d020f81f1d0374 cannot receive authenticated message: not established, status=0
2021-04-12T14:18:59.044022Z -0.002 0   [info ] ../src/node/node_to_node.h:59        | Dropping invalid authenticated message from 1efdf2c85f19c26c0351b7881b272b87b152820e3cd822a2d9d020f81f1d0374
2021-04-12T14:18:59.044097Z -0.002 0   [debug] ../src/consensus/aft/raft.h:2158     | Recv request vote to d8c06b7581f77496f9c46e1b3fc76c0d4fb9e819e27beb12a4d70585fd5ab6ff from 1efdf2c85f19c26c0351b7881b272b87b152820e3cd822a2d9d020f81f1d0374: their term is later (3 < 4)
2021-04-12T14:18:59.044158Z -0.002 0   [trace] ../src/node/history.h:424            | mt_retract_to index=32
2021-04-12T14:18:59.044211Z -0.002 0   [debug] ../src/node/history.h:86             | History [rollback] <sha256 1f17777ecd9be7d81c48224a102b53b83049a063885961ae9f4ec1bb4c3725af>
2021-04-12T14:18:59.044265Z -0.002 0   [debug] ../src/consensus/aft/raft.h:2589     | Setting term in store to: 4
2021-04-12T14:18:59.044351Z        100 [debug] ../src/host/ledger.h:909             | Ledger truncate: 32/32
2021-04-12T14:18:59.044414Z -0.002 0   [debug] ../src/consensus/aft/raft.h:2592     | Rolled back at 32
2021-04-12T14:18:59.044514Z -0.002 0   [info ] ../src/consensus/aft/raft.h:2384     | Becoming follower d8c06b7581f77496f9c46e1b3fc76c0d4fb9e819e27beb12a4d70585fd5ab6ff: 4
```

Is this expected or should the node reject all requests for anything after becoming retired?

**Comment by jumaffre** [comment]
As discussed with @wintersteiger, this is because request vote messages are still received by the retired leader node even though the node-to-node channel has been closed. We have a fix for this to be rolled out in a PR that will close this issue.

I also tried to understand why other nodes sent request votes to the retired primary in the first place. This is because even though the retirement of the old primary has been observed by other nodes, it hasn't yet been committed by the network and the old primary still exists in the set of known nodes. The primary's retirement commits on the retired primary first, but doesn't inform the other nodes because it is now in state `Retired` and no longer sends append entries to them. So the other nodes (now candidates) still include the retired primary in their majority calculation. 

For example, if there were 4 nodes {0, 1, 2, 3} (quorum is 3) and the leader 0 is retired, remaining nodes {1,2,3} will want to see votes from all remaining nodes (election quorum is still 3 even though the real quorum is now 2) to elect a new leader (as if leader 0 has died instead of being retired). I believe this could become an issue if we'd retired two nodes (primary + one backup) at the same time (e.g. during a code upgrade) as the remaining nodes would want to wait for a vote response from the retired primary or backup to complete the election. Am I missing something?
